File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/I_hotel/hotel.py

RESERVE = 'R'
COUNT = 'A'

n,q = map(int, input().split())

BIT = [0]*(n+1)
avail = [1] * (n+1)

for _ in range(q):
   line = list(x for x in input().split())
   if line[0] == RESERVE:
      index = int(line[1])
      if not avail[index]: continue
      avail[index] = 0
      while index < len(BIT):
         BIT[index] += 1
         index += index-(index&(index-1))
   else:
      st = int(line[1])-1
      en = int(line[2])
      ans = en - st
      while st:
         ans += BIT[st]
         st -= st-(st&(st-1))
      while en:
         ans -= BIT[en]
         en -= en-(en&(en-1))
      print(ans)

File: RPC/RPC_10_12_10_2024/Soluciones/ThirteenthAnnual_UCF_LocalProgrammingContest/local2024FinalRoundData/solutions/I_hotel/hotel.java
// Arup Guha
// 8/14/2024
// Solution to 2024 UCF Locals Problem: Hotel Rooms

import java.util.*;
import java.io.*;

public class hotel {

	public static void main(String[] args) throws Exception {
	
		BufferedReader stdin = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer tok = new StringTokenizer(stdin.readLine());
		int n = Integer.parseInt(tok.nextToken());
		int numOps = Integer.parseInt(tok.nextToken());
		
		// Build answer here.
		StringBuffer sb = new StringBuffer();
		
		// Stores hotel contents.
		bit ramada = new bit(n);
		
		// Process operations.
		for (int i=0; i<numOps; i++) {
			
			tok = new StringTokenizer(stdin.readLine());
			char cmd = tok.nextToken().charAt(0);
			
			// Reserve a new room.
			if (cmd == 'R') {
				int room = Integer.parseInt(tok.nextToken());
				ramada.add(room, 1);
			}
			
			// Query.
			else {
				int low = Integer.parseInt(tok.nextToken());
				int high = Integer.parseInt(tok.nextToken());
				
				// My bit stores taken rooms, so we would just subtract the taken
				// rooms from the total number of rooms in the range.
				sb.append((high - low + 1 - ramada.rangeSum(low,high))+"\n");
			}
		}
		
		// Ta da!
		System.out.print(sb);
	}
}

class bit {

	public int n;
	public int[] sum;
	
	public bit(int myn) {
	
		// Set size.
		n = 1;
		while (n < myn+1) n = (n<<1);
		
		// Allocate array.
		sum = new int[n];
	}
	
	// Add val to index idx.
	public void add(int idx, int val) {
		while (idx < n) {
			sum[idx] += val;
			idx += (idx&(-idx));
		}
	}
	
	// Returns the sum of arr[1]..arr[top];
	public int query(int top) {
		int res = 0;
		while (top > 0) {
			res += sum[top];
			top -= (top&(-top));
		}
		return res;
	}
	
	// Returns the sum arr[low]..arr[high]
	public int rangeSum(int low, int high) {
		return query(high) - query(low-1);
	}
}